File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        close_paren = {')': '(', '}': '{', ']': '['}

        for ch in s:
            if ch in close_paren.values():  # if ch is an opening bracket
                stack.append(ch)
            elif ch in close_paren:  # if ch is a closing bracket
                top_element = stack.pop() if stack else '#'
                if close_paren[ch] != top_element:
                    return False
            else:
                return False  # invalid character

        return not stack  # stack should be empty if all brackets are balanced

File: 0020-valid-parentheses/0020-valid-parentheses.ts
function isValid(s: string): boolean {
    let stack = []
    const bracked_map = { ")": "(", "]":"[", "}":"{" }

    for (let char of s) {
        if ("({[".includes(char)) {
            stack.push(char)
        } else if (stack && bracked_map[char] === stack[stack.length-1]){
            stack.pop()
        } else {
            return false
        }
    }

    return stack.length === 0
};